In [4]:
import pandas as pd
import warnings

warnings.filterwarnings("ignore")

In [8]:
data = pd.read_csv(r'C:\Priyanka\job_application_2024\JPMorganChase\Take Home Project\training_loan_data.csv', header=1)
data.head(5)

id   member_id  loan_amnt        term int_rate emp_length  \
0  10000001  11983056.0       7550   36 months   16.24%    3 years   
1  10000002  12002921.0      27050   36 months   10.99%  10+ years   
2  10000003  11983096.0      12000   36 months   10.99%    4 years   
3  10000004  12003142.0      28000   36 months    7.62%    5 years   
4  10000005  11993233.0      12000   36 months   13.53%  10+ years   

  home_ownership  annual_inc  \
0           RENT     28000.0   
1            OWN     55000.0   
2           RENT     60000.0   
3       MORTGAGE    325000.0   
4           RENT     40000.0   

                                                desc             purpose  ...  \
0                                                NaN  debt_consolidation  ...   
1    Borrower added on 12/31/13 > Combining high ...  debt_consolidation  ...   
2    Borrower added on 12/31/13 > I would like to...  debt_consolidation  ...   
3                                                NaN  debt_consolidation  ...   
4                                                NaN  debt_consolidation  ...   

   inq_last_6mths  mths_since_recent_inq  revol_util  total_bc_limit  \
0             0.0                   17.0         72%          4000.0   
1             0.0                    8.0      61.20%         35700.0   
2             1.0                    3.0         24%         18100.0   
3             1.0                    3.0      54.60%         42200.0   
4             0.0                   17.0      68.80%          7000.0   

   mths_since_last_major_derog tot_hi_cred_lim  tot_cur_bal  \
0                          NaN     3828.953801       5759.0   
1                          NaN    34359.940730     114834.0   
2                          NaN    16416.617760       7137.0   
3                          NaN    38014.149760     799592.0   
4                         53.0     6471.462236      13605.0   

   application_approved_flag  internal_score  bad_flag  
0                          1              99       0.0  
1                          1             353       0.0  
2                          1             157       0.0  
3                          1             365       0.0  
4                          1             157       0.0  

[5 rows x 23 columns]

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199121 entries, 0 to 199120
Data columns (total 23 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   id                           199121 non-null  int64  
 1   member_id                    189457 non-null  float64
 2   loan_amnt                    199121 non-null  int64  
 3   term                         189457 non-null  object 
 4   int_rate                     189457 non-null  object 
 5   emp_length                   181531 non-null  object 
 6   home_ownership               189457 non-null  object 
 7   annual_inc                   189457 non-null  float64
 8   desc                         82004 non-null   object 
 9   purpose                      189457 non-null  object 
 10  percent_bc_gt_75             180419 non-null  float64
 11  bc_util                      180333 non-null  float64
 12  dti                          189457 non-null  float64
 13 

In [13]:
data.describe(include='all').T

count unique  \
id                           199121.0    NaN   
member_id                    189457.0    NaN   
loan_amnt                    199121.0    NaN   
term                           189457      2   
int_rate                       189457    145   
emp_length                     181531     11   
home_ownership                 189457      5   
annual_inc                   189457.0    NaN   
desc                            82004  80643   
purpose                        189457     13   
percent_bc_gt_75             180419.0    NaN   
bc_util                      180333.0    NaN   
dti                          189457.0    NaN   
inq_last_6mths               189457.0    NaN   
mths_since_recent_inq        161472.0    NaN   
revol_util                     189330   1068   
total_bc_limit               181962.0    NaN   
mths_since_last_major_derog   32749.0    NaN   
tot_hi_cred_lim              181962.0    NaN   
tot_cur_bal                  161716.0    NaN   
application_approved_flag    199121.0    NaN   
internal_score               199121.0    NaN   
bad_flag                     189457.0    NaN   

                                                                           top  \
id                                                                         NaN   
member_id                                                                  NaN   
loan_amnt                                                                  NaN   
term                                                                 36 months   
int_rate                                                                12.12%   
emp_length                                                           10+ years   
home_ownership                                                        MORTGAGE   
annual_inc                                                                 NaN   
desc                           Borrower added on 09/05/13 > Debt consolidat...   
purpose                                                     debt_consolidation   
percent_bc_gt_75                                                           NaN   
bc_util                                                                    NaN   
dti                                                                        NaN   
inq_last_6mths                                                             NaN   
mths_since_recent_inq                                                      NaN   
revol_util                                                                  0%   
total_bc_limit                                                             NaN   
mths_since_last_major_derog                                                NaN   
tot_hi_cred_lim                                                            NaN   
tot_cur_bal                                                                NaN   
application_approved_flag                                                  NaN   
internal_score                                                             NaN   
bad_flag                                                                   NaN   

                               freq             mean             std  \
id                              NaN  10098566.911205    57043.972099   
member_id                       NaN   5929625.920367  3339821.824345   
loan_amnt                       NaN     14071.947334     8067.076636   
term                         144800              NaN             NaN   
int_rate                       9408              NaN             NaN   
emp_length                    61648              NaN             NaN   
home_ownership                97647              NaN             NaN   
annual_inc                      NaN     72243.300905    51804.829944   
desc                              9              NaN             NaN   
purpose                      112259              NaN             NaN   
percent_bc_gt_75                NaN        53.575807       34.153946   
bc_util                         NaN        66.

### Above table gives us initial observation of the data.
- member_id is not unique in the dataset. There could be the reason of duplicate records or these members have taken multiple loans. We can confirm duplicate records and remove them.
- We wont need id and member_id columns for model training
- term column has only 2 categories, we can just create replace with 0 or 1 based on category
- int_rate and emp_length are shown as categorical columns with 145 and 11 categories respectively.We can analyze further and change it to numerical columns or use techniques like binning to combine into few categories
- desc column has >50% missing data as there are ~82K records filled with desc out of ~199k records.
- bad_flag (target variable) has mean of ~0.06 and std of ~0.25. data is right skewed. This tells us that data is imbalanced. We can try difference oversampling/undersampling techniques later during model training. 

### EDA and Feature Engineering

##### Check and remove the duplicate records

In [ ]:
# Find total number of duplicate records in the dataset
data.duplicated().sum()

1334

In [21]:
data[data.duplicated()==True].head(5)

id  member_id  loan_amnt        term int_rate emp_length  \
197787  10049375  8736263.0      16000   36 months   11.55%  10+ years   
197788  10049376  8736163.0       4950   36 months   21.15%  10+ years   
197789  10049377  8453521.0      20000   36 months   13.05%    3 years   
197790  10049378  8746108.0      15000   36 months    6.03%  10+ years   
197791  10049379  8736072.0       8000   36 months    7.90%    5 years   

       home_ownership  annual_inc  \
197787           RENT     70777.0   
197788           RENT     32000.0   
197789       MORTGAGE    135000.0   
197790       MORTGAGE     75000.0   
197791           RENT     39000.0   

                                                     desc             purpose  \
197787                                                NaN  debt_consolidation   
197788                                                NaN               other   
197789                                                NaN  debt_consolidation   
197790                                                NaN  debt_consolidation   
197791    Borrower added on 09/08/13 > pay off other l...  debt_consolidation   

        ...  inq_last_6mths  mths_since_recent_inq  revol_util  \
197787  ...             0.0                   22.0      60.60%   
197788  ...             0.0                    NaN      69.10%   
197789  ...             2.0                    1.0      51.80%   
197790  ...             0.0                   11.0      26.20%   
197791  ...             1.0                    3.0      22.50%   

        total_bc_limit  mths_since_last_major_derog tot_hi_cred_lim  \
197787         25300.0                          NaN     25182.18502   
197788          2500.0                          NaN      2663.83748   
197789         16600.0                          NaN     16868.55070   
197790         26300.0                          NaN     28309.65693   
197791         88700.0                          NaN     89411.32780   

        tot_cur_bal  application_approved_flag  internal_score  bad_flag  
197787      43999.0                          1             209       0.0  
197788       7958.0                          1              65       0.0  
197789     190590.0                          1             261       0.0  
197790     538220.0                          1             196       0.0  
197791      29992.0                          1             105       0.0  

[5 rows x 23 columns]

In [22]:
# Confirming if id = 10049375 has multiple entries. 
data[data['id']==10049375]

id  member_id  loan_amnt        term int_rate emp_length  \
49374   10049375  8736263.0      16000   36 months   11.55%  10+ years   
197787  10049375  8736263.0      16000   36 months   11.55%  10+ years   

       home_ownership  annual_inc desc             purpose  ...  \
49374            RENT     70777.0  NaN  debt_consolidation  ...   
197787           RENT     70777.0  NaN  debt_consolidation  ...   

        inq_last_6mths  mths_since_recent_inq  revol_util  total_bc_limit  \
49374              0.0                   22.0      60.60%         25300.0   
197787             0.0                   22.0      60.60%         25300.0   

        mths_since_last_major_derog tot_hi_cred_lim  tot_cur_bal  \
49374                           NaN     25182.18502      43999.0   
197787                          NaN     25182.18502      43999.0   

        application_approved_flag  internal_score  bad_flag  
49374                           1             209       0.0  
197787                          1             209       0.0  

[2 rows x 23 columns]

In [23]:
# Keep the first occurrence of the records and remove duplicates. 
data = data.drop_duplicates(keep='first')

#Confirming if duplicates are dropped or not
print(data.duplicated().sum())
print(data[data['id']==10049375])

0
             id  member_id  loan_amnt        term int_rate emp_length  \
49374  10049375  8736263.0      16000   36 months   11.55%  10+ years   

      home_ownership  annual_inc desc             purpose  ...  \
49374           RENT     70777.0  NaN  debt_consolidation  ...   

       inq_last_6mths  mths_since_recent_inq  revol_util  total_bc_limit  \
49374             0.0                   22.0      60.60%         25300.0   

       mths_since_last_major_derog tot_hi_cred_lim  tot_cur_bal  \
49374                          NaN     25182.18502      43999.0   

       application_approved_flag  internal_score  bad_flag  
49374                          1             209       0.0  

[1 rows x 23 columns]
